tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,114 @@
+        this.storeMailboxManager = storeMailboxManager;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash(RunningOptions runningOptions) {
+        return Flux.from(usersRepository.listReactive())
+            .transform(ReactorUtils.<Username, Task.Result>throttle()
+                .elements(runningOptions.getUsersPerSecond())
+                .per(Duration.ofSeconds(1))
+                .forOperation(this::cleanupTrashForSingleUser))
+            .reduce(Task.Result.COMPLETED, Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while accessing users from repository", e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username) {
+        return Mono.from(jmapSettingsRepository.get(username))
+            .flatMap(jmapSettings -> cleanupTrashForSingleUser(username, jmapSettings))
+            .defaultIfEmpty(Task.Result.COMPLETED)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while cleaning trash for user {}", username, e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username, JmapSettings jmapSettings) {
+        if (jmapSettings.trashCleanupEnabled()) {
+            return cleanupTrashForSingleUserWithSpecificPeriod(username,
+                mapPeriodSettingToDuration.getOrDefault(jmapSettings.trashCleanupSetting(), Period.ofMonths(1)));
+        } else {
+            return Mono.just(Task.Result.COMPLETED);
+        }
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUserWithSpecificPeriod(Username username, Period period) {
+        return getMessageManagerForTrashMailbox(username)
+            .flatMap(messageManager ->
+                deleteExpiredMessages(storeMailboxManager.getSessionProvider().createSystemSession(username),
+                    getExpiredDate(period),
+                    messageManager))
+            .then(Mono.just(Task.Result.COMPLETED));
+    }
+
+    private Mono<MessageManager> getMessageManagerForTrashMailbox(Username username) {
+        return Mono.from(systemMailboxesProvider.getMailboxByRole(Role.TRASH, username));
+    }
+
+    private Mono<Void> deleteExpiredMessages(MailboxSession mailboxSession, Instant expiredDate, MessageManager messageManager) {
+        return Flux.from(messageManager.getMessagesReactive(MessageRange.all(), FetchGroup.MINIMAL, mailboxSession))
+            .filter(messageResult -> messageResult.getSaveDate().orElse(messageResult.getInternalDate()).toInstant().isBefore(expiredDate))
+            .map(MessageResult::getUid)
+            .collect(ImmutableList.toImmutableList())
+            .flatMap(messageUids -> messageManager.deleteReactive(messageUids, mailboxSession));
```

Comment:
should we worry about bulk deleting all messageUids when they have big sizes?

I see the same worry from Benoit at archival task https://github.com/linagora/tmail-backend/pull/838#issuecomment-1745153435

Reply:
That escaped me actually... I think you are correct. We should have a param in runningOptions that is messagesPerSecond and not usersPerSecond, do the throttling on messages and not users (concurrency of 1 by default).

I see similar patterm on other tasks

Reply:
likely should use `Flux.window` to batch deletion then.

Reply:
Didn't know that one... starts sounding like dark reactor magic but ok :)

Reply:
`ReactorUtils::throttle` is basically based on `Flux.window` :p

Reply:
Sorry half burned brain here :) yes yes throttle on the deletion, that's what I meant above. We agree!

Reply:
Well checking the underlying storage code, I see the `deleteReactive` API is already handling the concurrency issue: https://github.com/apache/james-project/blob/master/mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java#L336

So likely do not need to care about it here.

Reply:
it looks good
```
chunk.size.expunge
Optional. Defaults to 50.
Controls the number of messages to be expunged in parallel.
```